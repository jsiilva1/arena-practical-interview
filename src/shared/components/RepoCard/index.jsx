/**
 * External Dependencies
 */
import React from 'react';

/**
 * Internal Dependencies
 */
import { breakWords } from '../../utils/helpers';
import {
  CardItem,
  CardBody,
  Stats,
  StatItem,
  User,
  UserPhoto,
  UserInfo,
} from './styles';
import Icon from '../Icon';

const RepoCard = (props) => {
  const {
    title,
    description,
    forksNumber,
    starsNumber,
    user,
  } = props;

  const {
    username,
    type,
    avatar_url,
  } = user;

  return (
    <CardItem title="View repo details">
      <CardBody>
        <h1 className="list-title">{title}</h1>
        <h2 className="description">{breakWords({ text: description, toCharLimit: 85 })}</h2>

        <Stats>
          <StatItem title={`${forksNumber} forks`}>
            <Icon name={['fas', 'code-branch']} vendor="fa" />
            <span className="number-stat">{forksNumber}</span>
          </StatItem>

          <StatItem title={`${starsNumber} stars`}>
            <Icon name={['fas', 'star']} vendor="fa" />
            <span className="number-stat">{starsNumber}</span>
          </StatItem>
        </Stats>

        <User>
          <UserPhoto>
            <img src={avatar_url} alt={username} />
          </UserPhoto>

          <UserInfo>
            <p className="user-fullname">{username}</p>
            <p className="user-username">{type}</p>
          </UserInfo>
        </User>

      </CardBody>
    </CardItem>
  );
};

export default RepoCard;
