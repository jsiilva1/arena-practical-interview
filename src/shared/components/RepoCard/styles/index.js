import styled from 'styled-components';

import { device } from '../../../theme/settings/screens';

export const CardWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
`;

export const CardItem = styled.div`
  background: #FFF;
  width: calc(80vh - 21%);
  min-width: 250px;
  box-shadow: rgba(0, 0, 0, 0.04) 0px 3px 4px;
  border: 1px solid #ecf0f1;
  border-radius: 7px;
  margin: 0px 7px 7px 0;
  transition: 0.8s;
  cursor: pointer;
  position: relative;

  &:hover { opacity: 0.7; }
  
  @media ${device.tablet}, ${device.mobile} {
    .list-title {
      font-size: 1.5rem;
    }
  }

  @media ${device.mobile} {
    width: 100%;
    margin: 0px 0 10px 0;
    
    &:last-child {
      margin-bottom: 26%;
    }
  }

  & .list-title {
    margin: 0;
    font-weight: 800;
    font-style: bold;
    font-size: 1.4rem;
  }
  
  & .description {
    margin: 5px 0;
    font-size: 0.98rem;
    font-weight: 500;
    font-style: normal;
    color: #B8B8B8;
    max-height: 100px;
  }
`;

export const Stats = styled.div`
  margin-top: 10px;
  font-size: 0.82rem;
`;

export const StatItem = styled.span`
  margin: 0 6px;

  &:first-child {
    margin: 0;
  }

  & .number-stat {
    margin-left: 3px;
  }

  cursor: default;
`;

export const User = styled.span`
  width: 100%; 
  border-top: 1px solid #ecf0f1;
  float: left;
  margin: 20px 0 10px 0;
  padding-top: 10px;
`;

export const UserPhoto = styled.span` 
  & img { 
    width: 30px; 
    height: 30px; 
  }

  float: left;
`;

export const UserInfo = styled.div`
  display: flex;
  flex-direction: column;
  font-size: 0.7rem;

  & p {
    margin: 0 10px;
  }

  & p:first-child {
    margin-top: 2px;
  }

  & > .user-fullname { font-weight: 800; }
`;

export const CardHeader = styled.div`
  border-bottom: 1px solid #D0D0D0;
`;

export const CardBody = styled.div`
  margin: 20px;
`;
