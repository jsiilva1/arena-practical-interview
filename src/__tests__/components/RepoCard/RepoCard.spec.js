import React from 'react';
import { shallow } from 'enzyme';

import RepoCard from '../../../shared/components/RepoCard';

describe('RepoCard Component', () => {
  it('should render a Repo Card correctly', () => {
    const wrapper = shallow(
      <RepoCard
        key={1}
        title="react"
        description="A declarative, efficient, and flexible JavaScript library for building user interfaces."
        starsNumber={141000}
        forksNumber={27000}
        user={{ username: 'facebook', type: 'Organization', avatar_url: 'https://avatars2.githubusercontent.com/u/69631?s=200&v=4' }}
      />,
    );
    
    expect(wrapper).toMatchSnapshot();
  });
});
