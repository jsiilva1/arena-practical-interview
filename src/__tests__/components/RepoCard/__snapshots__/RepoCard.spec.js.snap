// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepoCard Component should render a Repo Card correctly 1`] = `
<styled.div
  title="View repo details"
>
  <styled.div>
    <h1
      className="list-title"
    >
      react
    </h1>
    <h2
      className="description"
    >
      A declarative, efficient, and flexible JavaScript library for building user interface...
    </h2>
    <styled.div>
      <styled.span
        title="27000 forks"
      >
        <Icon
          name={
            Array [
              "fas",
              "code-branch",
            ]
          }
          vendor="fa"
        />
        <span
          className="number-stat"
        >
          27.0k
        </span>
      </styled.span>
      <styled.span
        title="141000 stars"
      >
        <Icon
          name={
            Array [
              "fas",
              "star",
            ]
          }
          vendor="fa"
        />
        <span
          className="number-stat"
        >
          141.0k
        </span>
      </styled.span>
    </styled.div>
    <styled.span>
      <styled.span>
        <img
          alt="facebook"
          src="https://avatars2.githubusercontent.com/u/69631?s=200&v=4"
        />
      </styled.span>
      <styled.div>
        <p
          className="user-fullname"
        >
          facebook
        </p>
        <p
          className="user-username"
        >
          Organization
        </p>
      </styled.div>
    </styled.span>
  </styled.div>
</styled.div>
`;
